PromptForChoice Errors in CI Pipelines

## Description

I get errors when running CI in Github actions for code that is using AutomatedLab.

Read-Choice : Exception calling "PromptForChoice" with "4" argument(s): "Object reference not set to an instance of an 
object."

## Expected behavior

It would be great to have some sort of `Unattend` switch or `AUTOMATEDLAB_UNATTEND` environment variable for taking default instead of keyboard input.

## Actual behavior

```output
[19:12:07][AutomatedLab.init.ps1] Your version of AutomatedLab is outdated. Consider updating to the recent version, 5.41.0
C:\LabSources
C:\LabSources\ISOs
19:14:55|00:00:00|00:00:00.000| Initialization
19:14:55|00:00:00|00:00:00.050| - Host operating system version: 'Microsoft Windows Server 2022 Datacenter, 10.0.20348.0'
19:14:55|00:00:00|00:00:00.057| - Creating new lab definition with name 'SPDev'
19:14:55|00:00:00|00:00:00.070| - Location of lab definition files will be 'C:\ProgramData\AutomatedLab/Labs/SPDev'
19:14:55|00:00:00|00:00:00.000| - Performing update of SysInternals suite and lab sources directory now
19:16:18|00:01:23|00:01:22.503|   - SysInternals Suite has been updated and placed in 'C:\LabSources\Tools\SysInternals'
19:16:18|00:01:23|00:01:22.601| - Location of LabSources folder is 'C:\LabSources'
19:16:18|00:01:23|00:00:00.000| - Adding Azure subscription data
19:16:20|00:01:[25](https://github.com/alexsapozhkov/SharePointDeployment/runs/5308825987?check_suite_focus=true#step:4:25)|00:00:01.524|   - Using Azure Subscription 'Visual Studio Enterprise – MPN' (***)
19:16:27|00:01:32|00:00:08.844|   - Using Azure Location 'West Europe'
19:16:27|00:01:32|00:00:08.851|   - Trying to locate or create default resource group
19:16:30|00:01:35|00:00:11.724|   - Querying available vm sizes for Azure location 'West Europe'
19:17:14|00:02:19|00:00:00.000|   - Looking for Azure LabSources inside subscription 'Visual Studio Enterprise – MPN'
19:17:18|00:02:22|00:00:03.[26](https://github.com/alexsapozhkov/SharePointDeployment/runs/5308825987?check_suite_focus=true#step:4:26)3|     - Azure LabSources verified / created
19:17:18|00:02:22|00:00:00.001|   - Auto-adding ISO files from Azure labsources share
19:17:[30](https://github.com/alexsapozhkov/SharePointDeployment/runs/5308825987?check_suite_focus=true#step:4:30)|00:02:[35](https://github.com/alexsapozhkov/SharePointDeployment/runs/5308825987?check_suite_focus=true#step:4:35)|00:00:12.490|     - No ISO files have been found in your Azure labsources share. Please make sure that they are present when you try mounting them.
19:17:30|00:02:35|00:00:12.497|     - Done
Read-Choice : Exception calling "PromptForChoice" with "4" argument(s): "Object reference not set to an instance of an 
object."
At C:\Program Files\WindowsPowerShell\Modules\AutomatedLab\5.40.0\AutomatedLabAzure.psm1:[38](https://github.com/alexsapozhkov/SharePointDeployment/runs/5308825987?check_suite_focus=true#step:4:38)1 char:13
+             Read-Choice -ChoiceList '&Yes', '&No, do not ask me again ...
+             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Read-Choice], MethodInvocationException
    + FullyQualifiedErrorId : NullReferenceException,Read-Choice
 
Error: Process completed with exit code 1.
```

https://github.com/alexsapozhkov/SharePointDeployment/runs/5308825987?check_suite_focus=true

## Steps to reproduce

```powershell
$labName = 'SPDev'
$domainName = 'contoso.com'
$vitualizationEngine = 'Azure'
$azureDefaultLocation = 'West Europe'
$password = 'MySecretPassword!'

# Automation: Configure telemetry opt-in or opt-out
[Environment]::SetEnvironmentVariable('AUTOMATEDLAB_TELEMETRY_OPTIN', 'yes', 'Machine')
$env:AUTOMATEDLAB_TELEMETRY_OPTIN = 'true'
Import-Module AutomatedLab

if ((Test-Path -Path 'C:\LabSources') -eq $false)
{
    New-LabSourcesFolder -Drive C -Force
}

# Copy all required ISO files to
"$labsources\ISOs" # $labSources is a dynamic variable and will point to lab sources

# Create new AL Lab definition
New-LabDefinition -Name $labname -DefaultVirtualizationEngine $vitualizationEngine

if ($vitualizationEngine -eq 'Azure')
{
    # Check if logged into Azure, else logon
    $context = Get-AzContext
    if ($null -eq $context.Account)
    {
        $null = Login-AzAccount
    }

    # Add current Azure subscription to new lab definition
    Add-LabAzureSubscription -DefaultLocationName $azureDefaultLocation
}
```

Here's the job definition in Github Actions:

```
name: CI
on: [push]
jobs:
  SmallNonHAMinrole-Binaries-SE-Win2022:
    runs-on: windows-2022
    steps:
      - uses: actions/checkout@v2
      - name: Install Prerequisites
        shell: powershell
        run: |
          .\InstallPrerequisites.ps1
      - name: Provision Lab
        shell: powershell
        env:
          ARM_CLIENT_ID: ${{ secrets.ARM_CLIENT_ID }}
          ARM_CLIENT_SECRET: ${{ secrets.ARM_CLIENT_SECRET }}
          ARM_SUBSCRIPTION_ID: ${{ secrets.ARM_SUBSCRIPTION_ID }}
          ARM_TENANT_ID: ${{ secrets.ARM_TENANT_ID }}
        run: |
          $securedPassword = ConvertTo-SecureString $env:ARM_CLIENT_SECRET -AsPlainText -Force;
          $azureCredential = New-Object System.Management.Automation.PSCredential( $env:ARM_CLIENT_ID, $securedPassword );
          Connect-AzAccount -Credential $azureCredential -ServicePrincipal -TenantId $env:ARM_TENANT_ID | Out-Null;
          Select-AzSubscription -SubscriptionId $env:ARM_SUBSCRIPTION_ID;
          .\DeploySharePointEnvironment.ps1
```

## Environment details

PowerShell Version <!-- Required -->
```output
shell: C:\Windows\System32\WindowsPowerShell\v1.0\powershell.EXE -command ". '{0}'"
Name                           Value                                                                                   
----                           -----                                                                                   
PSVersion                      5.1.2034[8](https://github.com/alexsapozhkov/SharePointDeployment/runs/5310437697?check_suite_focus=true#step:3:8).320                                                                           
PSEdition                      Desktop                                                                                 
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                 
BuildVersion                   10.0.20348.320                                                                          
CLRVersion                     4.0.3031[9](https://github.com/alexsapozhkov/SharePointDeployment/runs/5310437697?check_suite_focus=true#step:3:9).42000                                                                         
WSManStackVersion              3.0                                                                                     
PSRemotingProtocolVersion      2.3                                                                                     
SerializationVersion           1.1.0.1
```

AutomatedLab components <!-- Required -->
```output
Name    : AutomatedLab
Version : 5.40.0
Name    : AutomatedLab.Common
Version : 2.1.[22](https://github.com/alexsapozhkov/SharePointDeployment/runs/5310465077?check_suite_focus=true#step:3:22)3
Name    : AutomatedLabDefinition
Version : 5.41.0
Name    : AutomatedLabNotifications
Version : 5.41.0
Name    : AutomatedLabTest
Version : 5.41.0
Name    : AutomatedLabUnattended
Version : 5.41.0
Name    : AutomatedLabWorker
Version : 5.41.0
```

--- a/AutomatedLab/AutomatedLab.init.ps1
+++ b/AutomatedLab/AutomatedLab.init.ps1
@@ -118,6 +118,7 @@ else
 Set-PSFConfig -Module 'AutomatedLab' -Name OsRoot -Value $osroot -Initialize -Validation string
 Set-PSFConfig -Module 'AutomatedLab' -Name OverridePowerPlan -Value $true -Initialize -Validation bool -Description 'On Windows: Indicates that power settings will be set to High Power during lab deployment'
 Set-PSFConfig -Module 'AutomatedLab' -Name SendFunctionTelemetry -Value $false -Initialize -Validation bool -Description 'Indicates if function call telemetry is sent' -Hidden
+Set-PSFConfig -Module 'AutomatedLab' -Name DoNotPrompt -Value $false -Initialize -Validation bool -Description 'Indicates that AutomatedLab should not display prompts. Workaround for environments that register as interactive, even if they are not' -Hidden
 
 #PSSession settings
 Set-PSFConfig -Module 'AutomatedLab' -Name InvokeLabCommandRetries -Value 3 -Initialize -Validation integer -Description 'Number of retries for Invoke-LabCommand'

--- a/AutomatedLab/AutomatedLab.psm1
+++ b/AutomatedLab/AutomatedLab.psm1
@@ -513,7 +513,11 @@ function Import-Lab
                     {
                         $Path = Join-Path -Path (Get-Lab).Sources.UnattendedXml.Value -ChildPath 'Unattended2012.xml'
                     }
-                    if ($this.OperatingSystemType -eq 'Linux' -and $this.LinuxType -eq 'RedHat')
+                    if ($this.OperatingSystemType -eq 'Linux' -and $this.LinuxType -eq 'RedHat' -and $this.Version -lt 8.0)
+                    {
+                        $Path = Join-Path -Path (Get-Lab).Sources.UnattendedXml.Value -ChildPath ks_defaultLegacy.cfg
+                    }
+                    if ($this.OperatingSystemType -eq 'Linux' -and $this.LinuxType -eq 'RedHat' -and $this.Version -ge 8.0)
                     {
                         $Path = Join-Path -Path (Get-Lab).Sources.UnattendedXml.Value -ChildPath ks_default.cfg
                     }

--- a/AutomatedLab/AutomatedLabAzure.psm1
+++ b/AutomatedLab/AutomatedLabAzure.psm1
@@ -393,7 +393,7 @@ function Add-LabAzureSubscription
     $syncMaxSize = Get-LabConfigurationItem -Name LabSourcesMaxFileSizeMb
     $syncIntervalDays = Get-LabConfigurationItem -Name LabSourcesSyncIntervalDays
 
-    if (-not $lastchecked)
+    if (-not (Get-LabConfigurationItem -Name DoNotPrompt -Default $false) -and -not $lastchecked)
     {
         $lastchecked = [datetime]0
         $syncText = @"

--- a/AutomatedLab/AutomatedLabVirtualMachines.psm1
+++ b/AutomatedLab/AutomatedLabVirtualMachines.psm1
@@ -1240,9 +1240,10 @@ function Connect-LabVM
 
         if ($machine.OperatingSystemType -eq 'Linux')
         {
-            $sshBinary = Get-ChildItem $labsources\Tools\OpenSSH -Filter ssh.exe -Recurse -ErrorAction SilentlyContinue | Select-Object -First 1
+            $sshBinary = Get-Command ssh.exe -ErrorAction SilentlyContinue
+            if (-not $sshBinary) { Get-ChildItem $labsources\Tools\OpenSSH -Filter ssh.exe -Recurse -ErrorAction SilentlyContinue | Select-Object -First 1 }
 
-            if (-not $sshBinary)
+            if (-not $sshBinary -and -not (Get-LabConfigurationItem -Name DoNotPrompt))
             {
                 $download = Read-Choice -ChoiceList 'No','Yes' -Caption 'Download Win32-OpenSSH' -Message 'OpenSSH is necessary to connect to Linux VMs. Would you like us to download Win32-OpenSSH for you?' -Default 1
 

--- a/AutomatedLabDefinition/AutomatedLabDefinition.psm1
+++ b/AutomatedLabDefinition/AutomatedLabDefinition.psm1
@@ -1245,6 +1245,7 @@ function Export-LabDefinition
             if ($Script:machines | Where-Object LinuxType -eq 'RedHat')
             {
                 $kickstartContent | Out-File -FilePath (Join-Path -Path $script:lab.Sources.UnattendedXml.Value -ChildPath ks_default.cfg) -Encoding unicode
+                $kickstartContent.Replace(' --non-interactive','') | Out-File -FilePath (Join-Path -Path $script:lab.Sources.UnattendedXml.Value -ChildPath ks_defaultLegacy.cfg) -Encoding unicode                
             }
             if ($Script:machines | Where-Object LinuxType -eq 'Suse')
             {
@@ -3626,7 +3627,11 @@ function Import-LabDefinition
                     {
                         $Path = Join-Path -Path (Get-Lab).Sources.UnattendedXml.Value -ChildPath 'Unattended2012.xml'
                     }
-                    if ($this.OperatingSystemType -eq 'Linux' -and $this.LinuxType -eq 'RedHat')
+                    if ($this.OperatingSystemType -eq 'Linux' -and $this.LinuxType -eq 'RedHat' -and $this.Version -lt 8.0)
+                    {
+                        $Path = Join-Path -Path (Get-Lab).Sources.UnattendedXml.Value -ChildPath ks_defaultLegacy.cfg
+                    }
+                    if ($this.OperatingSystemType -eq 'Linux' -and $this.LinuxType -eq 'RedHat' -and $this.Version -ge 8.0)
                     {
                         $Path = Join-Path -Path (Get-Lab).Sources.UnattendedXml.Value -ChildPath ks_default.cfg
                     }

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -12,6 +12,9 @@
   the function's scope of work.
 - Finally adding cluster-awareness
   - If AutomatedLab is executed locally on a cluster node, the lab VMs are added to the cluster
+- Adding hidden setting DoNotPrompt as an attempt to not prompt.
+  - Does not include prompts that can be turned off like Enable-LabHostRemoting -Force
+- Updated docs
 
 ### Bugs
 - Fixing issue with Get-LabAzureAvailableRoleSize by filtering earlier.
@@ -22,6 +25,7 @@
 - Fixes #1293: Set-DscLocalConfigurationManagerConfiguration throws error when successful.
 - DelayBetweenComputers was ignored
 - Fixed lab location. Sometimes the 'Lab.xml' was stored in 'C:\ProgramData\AutomatedLab' directly.
+- Removing unsupported kickstart parameter for old CentOS versions
 
 ## 5.41.0 (2022-01-31)
 

--- a/Help/AutomatedLab/en-us/Test-LabPathIsOnLabAzureLabSourcesStorage.md
+++ b/Help/AutomatedLab/en-us/Test-LabPathIsOnLabAzureLabSourcesStorage.md
@@ -23,10 +23,14 @@ Tests if a given path is inside the Azure lab source storage share
 
 ### Example 1
 ```powershell
-PS C:\> {{ Add example code here }}
+PS C:\> Test-LabPathIsOnLabAzureLabSourcesStorage -Path "$labSources\Tools\SomeTool.exe"
 ```
 
-{{ Add example description here }}
+Test if the path "$labSources\Tools\SomeTool.exe" is actually located on an Azure
+lab sources share.
+
+The dynamic $labSources variable may either point to a local path or an Azure
+file share depending on the Hypervisor that is used.
 
 ## PARAMETERS
 

--- a/Help/Wiki/Roles/dscpull.md
+++ b/Help/Wiki/Roles/dscpull.md
@@ -15,7 +15,7 @@ Add-LabMachineDefinition -Name Pulli -Roles DscPullServer
 You can optionally specify parameters by using the role definition:
 
 ```powershell
-$role = Get-LabMachineRoleDefinition -Role DscPullServer -RoleProperties @{
+$role = Get-LabMachineRoleDefinition -Role DscPullServer -Properties @{
     DatabaseEngine = 'mdb'
     DatabaseName = 'DSC'
     DatabaseServer = 'SQL01'

--- a/Help/Wiki/Roles/hyperv.md
+++ b/Help/Wiki/Roles/hyperv.md
@@ -13,7 +13,7 @@ Add-LabMachineDefinition -Name HV01 -Roles HyperV
 You can also customize many relevant settings by creating the role definition reference:
 
 ```powershell
-$role = Get-LabMachineRoleDefinition -Role HyperV -RoleProperties @{EnableEnhancedSessionMode = 'true'}
+$role = Get-LabMachineRoleDefinition -Role HyperV -Properties @{EnableEnhancedSessionMode = 'true'}
 Add-LabMachineDefinition -Name HV01 -Roles $role
 ```
 ## Role properties

--- a/Help/Wiki/Roles/sql.md
+++ b/Help/Wiki/Roles/sql.md
@@ -17,7 +17,7 @@ Add-LabMachineDefinition -Name SQL01 -Roles SQLServer2017
 If you need a little more control, simply use the role properties. The following example only installs the engine and tools:
 
 ```powershell
-$role = Get-LabMachineRoleDefinition -Role SQLServer2017 -RoleProperties @{Features = 'SQL,Tools'}
+$role = Get-LabMachineRoleDefinition -Role SQLServer2017 -Properties @{Features = 'SQL,Tools'}
 
 Add-LabMachineDefinition -Name SQL01 -Roles $role
 ```

